src/main/java/org/opensearch/knn/index/fetch/KNNFetchSubPhase.java

```diff
@@ -0,0 +1,235 @@
+        IndexSettings indexSettings = fetchContext.getIndexSettings();
+        if (!KNNSettings.isKNNSyntheticSourceEnabled(indexSettings)) {
+            log.debug("Synthetic is disabled for index: {}", fetchContext.getIndexName());
+            return null;
+        }
+        MapperService mapperService = fetchContext.mapperService();
+
+        List<DocValueField> fields = new ArrayList<>();
+        for (MappedFieldType mappedFieldType : mapperService.fieldTypes()) {
+            if (mappedFieldType != null && mappedFieldType instanceof KNNVectorFieldMapper.KNNVectorFieldType) {
+                String fieldName = mappedFieldType.name();
+                ValueFetcher fetcher = new DocValueFetcher(
+                    mappedFieldType.docValueFormat(null, null),
+                    fetchContext.searchLookup().doc().getForField(mappedFieldType)
+                );
+                fields.add(new DocValueField(fieldName, fetcher));
+            }
+        }
+        return new KNNFetchSubPhaseProcessor(fetchContext, fields);
+    }
+
+    @AllArgsConstructor
+    @Getter
+    class KNNFetchSubPhaseProcessor implements FetchSubPhaseProcessor {
+
+        private final FetchContext fetchContext;
+        private final List<DocValueField> fields;
+
+        @Override
+        public void setNextReader(LeafReaderContext readerContext) throws IOException {
+            for (DocValueField f : fields) {
+                f.fetcher.setNextReader(readerContext);
+            }
+        }
+
+        @Override
+        public void process(HitContext hitContext) throws IOException {
+            MapperService mapperService = fetchContext.mapperService();
+            final boolean hasNested = mapperService.hasNested();
+            SearchHit hit = hitContext.hit();
+            Map<String, Object> maps = hit.getSourceAsMap();
+            if (maps == null) {
+                // when source is disabled, return
+                return;
+            }
+
+            if (hasNested) {
+                syntheticNestedDocValues(mapperService, hitContext, maps);
+            }
+            for (DocValueField f : fields) {
+                if (maps.containsKey(f.field)) {
+                    continue;
+                }
+                List<Object> docValuesSource = f.fetcher.fetchValues(hitContext.sourceLookup());
+                if (docValuesSource.size() > 0) {
+                    maps.put(f.field, docValuesSource.get(0));
+                }
+            }
+            BytesStreamOutput streamOutput = new BytesStreamOutput(BYTES_PER_KILOBYTES);
+            XContentBuilder builder = new XContentBuilder(XContentType.JSON.xContent(), streamOutput);
+            builder.value(maps);
+            hitContext.hit().sourceRef(BytesReference.bytes(builder));
+        }
+
+        protected void syntheticNestedDocValues(MapperService mapperService, HitContext hitContext, Map<String, Object> maps)
+            throws IOException {
```

Comment:
Can we break down this method into several sub-methods? I am having some trouble following what it does. The rest of the PR looks good.

Also, does OpenSearch already have any utilities to do this?

Reply:
> Can we break down this method into several sub-methods? I am having some trouble following what it does. The rest of the PR looks good.

FIXED at https://github.com/opensearch-project/k-NN/pull/1571/commits/2a61fcda4ceef36829bbfa2a1748f2c5b6bbe7df

> Also, does OpenSearch already have any utilities to do this?

As i think KNNFetchSubPhase do not rely on OpenSearch, please correct me if I have some misunderstandings.